### PR TITLE
Add application level majorVersion override in application/v4 

### PR DESCRIPTION
**Pull request by bratseth**
We need to be able to override the applications override of the system level override of the system majorVersion on the application level.

@jonmv please review

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java [hunk 1]
```diff
@@ -182,60 +185,67 @@ public LockedApplication withoutDeploymentIn(ZoneId zone) {
...
 
     public LockedApplication withOwner(User owner) {
         return new LockedApplication(lock, id, createdAt, deploymentSpec, validationOverrides, deployments,
                 deploymentJobs, change, outstandingChange,
-                ownershipIssueId, Optional.ofNullable(owner), metrics, rotation, rotationStatus);
+                ownershipIssueId, Optional.ofNullable(owner), majorVersion, metrics, rotation, rotationStatus);
+    }
+
+    /** Set a major vewrsion for this, or set to null to remove any major version override */
+    public LockedApplication withMajorVersion(Integer majorVersion) {
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java]: Pretty sure I made a builder for `LockedApplication`. Where did it go 😱

Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java]: `vewrsion`.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/ApplicationList.java [hunk 2]
```diff
@@ -160,13 +160,13 @@ public ApplicationList canUpgradeAt(Instant instant) {
...
-     * Returns the subset of applications that hasn't pinned to an an earlier major version than the given one.
+     * Returns the subset of applications that hasn't pinned to another major version than the given one.
      *
      * @param targetMajorVersion the target major version which applications returned allows upgrading to
      * @param defaultMajorVersion the default major version to assume for applications not specifying one
      */
     public ApplicationList allowMajorVersion(int targetMajorVersion, int defaultMajorVersion) {
-        return listOf(list.stream().filter(a -> a.deploymentSpec().majorVersion().orElse(defaultMajorVersion)
+        return listOf(list.stream().filter(a -> a.majorVersion().orElse(a.deploymentSpec().majorVersion().orElse(defaultMajorVersion))
                                            >= targetMajorVersion));
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/ApplicationList.java]: Doc was in agreement with this, but not any more.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 3]
```diff
@@ -344,16 +352,28 @@ private HttpResponse applications(String tenantName, HttpRequest request) {
...
+        Inspector majorVersionField = toSlime(request.getData()).get().field("majorVersion");
+        if ( ! majorVersionField.valid())
+            throw new IllegalArgumentException("Request body must contain a majorVersion field");
+        Integer majorVersion = majorVersionField.asLong() == 0 ? null : (int)majorVersionField.asLong();
+        controller.applications().lockIfPresent(application.id(),
+                                                a -> controller.applications().store(a.withMajorVersion(majorVersion)));
+        return new MessageResponse("Set major version to " + ( majorVersion == null ? "empty" : majorVersion));
+    }
+
+    private Application getApplication(String tenantName, String applicationName, HttpRequest request) {
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: Can't find any reference to the `request` parameter in the method body.

**Review by jonmv — APPROVED** [review]
Perfect! A couple of nits, esp. about the ApplicationList method doc.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/ApplicationList.java (hunk 2 above)
Comment (bratseth) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/ApplicationList.java]: I think it's still correct

**Comment by bratseth** [comment]
Merge stop enforcer plz

**Comment by bratseth** [comment]
Merge stop enforcer common!

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/ApplicationList.java (hunk 2 above)
Comment (bratseth) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/ApplicationList.java]: Right, I forgot I edited this comment independently of the semantic change. I'll correct it back.

**Pull request merged**